presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/InternalResourceGroup.java

```diff
@@ -912,27 +1087,68 @@ public int hashCode()
         return Objects.hash(id);
     }
 
+    public InternalResourceGroup getRoot()
+    {
+        synchronized (root) {
+            return root;
+        }
+    }
+
+    public void clean()
+    {
+        synchronized (root) {
+            if (getRunningQueries() > 0 || getQueuedQueries() > 0) {
+                return;
+            }
+            for (InternalResourceGroup group : subGroups.values()) {
+                group.clean();
+            }
+            if (!getPrev().isPresent()) {
+                if (getRunningQueries() == 0 && getQueuedQueries() == 0) {
+                    if (next.isPresent() && parent.isPresent()) {
+                        next.get().setPrev(null);
+                        next = Optional.empty();
+                    }
+                    subGroups.clear();
+                    while (!eligibleSubGroups.isEmpty()) {
+                        eligibleSubGroups.poll();
+                    }
+                    dirtySubGroups.clear();
+                }
+            }
+        }
+    }
+
     @ThreadSafe
     public static final class RootInternalResourceGroup
             extends InternalResourceGroup
     {
-        public RootInternalResourceGroup(String name, BiConsumer<InternalResourceGroup, Boolean> jmxExportListener, Executor executor)
+        public RootInternalResourceGroup(String name, BiConsumer<InternalResourceGroup, Boolean> jmxExportListener, Executor executor, Integer version)
+        {
+            super(Optional.empty(), name, jmxExportListener, executor, version);
+        }
+
+        public RootInternalResourceGroup(String name, BiConsumer<InternalResourceGroup, Boolean> jmxExportListener, Executor executor, InternalResourceGroup root, Integer version)
```

Comment:
Why would a `RootInternalResourceGroup` take another resource group as root?

Reply:
The reason for is the root going to be the oldest version of the tree, which is the first root created for the given resource group tree. With new versions, i am also creating newer version of the root as that's how InternalResourceGroupManager#refreshAndStartQueries logic is based on. Find if any eligible query with older version, once done, check if the root has new version then search for eligible query to run with newer version.

Reply:
Isn't the older version accessible via `prev`? If it's really needed, at least name it and use it in the right way. Otherwise the definition of root group becomes unclear? I still feel that it's probably easier to manage to disallow changes to root group.